Contribution for the hyena chapter (#1)
lulmer/computer-vision-course: chapters/en/Unit 13 - Outlook/hyena.mdx

## Overview

### What is Hyena 

> Hyena is a new type of operator that substitute to the attention mechanism. Developed by Hazy Research, it features a subquadratic computational efficiency, constructed by interleaving implicitly parametrized long convolutions and data-controlled gating.

Reviewer: ```suggestion
Hyena is a new type of operator that substitutes the attention mechanism. Developed by Hazy Research, it features a subquadratic computational efficiency, constructed by interleaving implicitly parametrized long convolutions and data-controlled gating.
```

![transformer2hyena.png](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/outlook_hyena_images/transformer2hyena.png)

> It consists in recursively computing convolutions and multiplicative element-wise gating operations. This approach builds on top of the H3 mechanism, also developed by the same researchers. The H3 mechanism is characterized by its data-controlled, parametric decomposition, acting as a surrogate attention mechanism.

Reviewer: ```suggestion
It consists in recursively computing convolutions and multiplicative element-wise gating operations. This approach builds on top of the [Hungry Hungry Hippo (H3)](https://arxiv.org/abs/2212.14052) mechanism, also developed by the same researchers. The H3 mechanism is characterized by its data-controlled, parametric decomposition, acting as a surrogate attention mechanism.
```

The core idea is to repeatedly apply linear operators that are fast to evaluate to an input sequence $u \in \mathbb{R}^{L}$ with $L$ the length of the sequence. A notable design choice is the use of **implicit convolutions**. Unlike standard convolutional layers, the convolution filter $h$ is learned implicitly with a small neural network $\gamma_{\theta}$ (also called the Hyena Filter). This network takes the positional index and potentially positional encodings as inputs. From the outputs of $\gamma_{\theta}$ one can construct a Toeplitz matrix $T_{h}$. *This implies that instead of learning the values of the convolution filter directly, we learn a map from a temporal positional encoding to the values. Which is more computationally efficient, especially on long sequence regime. 

NB : It's important to note that the mapping function  $\gamma_{\theta}$  can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in H3 paper.

### Implicit convolutions

> Convolutions can be transformed into matrix multiplication using a Toeplitz matrix. 

Reviewer: ```suggestion
Convolutions can be transformed into matrix multiplication using a [Toeplitz matrix](https://en.wikipedia.org/wiki/Toeplitz_matrix). 
```

![toeplitz_conv.jpg](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/outlook_hyena_images/toeplitz_conv.jpg)
This transformation leads to greater parameter efficiency. Instead of directly learning fixed kernel weight values, a parametrized function is employed. This function intelligently deduces the values of the kernel weights and their dimensions during the network's forward pass, optimizing resource use.

In practice, convolutions are accelerated to a subquadratic time complexity by the Cooley-Tukey fast Fourier transform (FFT) algorithm. Some work has been conducted to speed up this computation like FastFFTConv based on Monarch decomposition. 
### Wrapping up everything

![nd_hyena.png](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/outlook_hyena_images/nd_hyena.png)
In essence, Hyena can be performed in two steps: 
1) Compute a set of N+1 linear projections similarly of attention (it can be more than 3 projections)
2) Mixing up the projections : The matrix H(u) is defined by a combination of matrix multiplications

## Why it matters

> While Transformer is a well established and very capable architecture, the quadratic computational cost is an expensive price to pay, specially in inference. The H3 mechanism proposition went close to the perplexity of multi-headed attention mechanisms, but there was still a narrow gap in terms of perplexity. 

Reviewer: ```suggestion
While Transformer is a well established and very capable architecture, the quadratic computational cost is an expensive price to pay, especially in inference. The H3 mechanism proposition went close to the perplexity of multi-headed attention mechanisms, but there was still a narrow gap in terms of perplexity. 
```

A variety of attention replacements have been proposed over the last few years, and it remains challenging to evaluate the quality of a new architecture during the exploratory phase. Creating a versatile layer that can effectively process N-Dimensional data within deep neural networks while keeping a good expressiveness is a significant area of ongoing research.

> Empirically, Hyena operators are able to significantly shrink the quality gap with attention at scale, reaching similar perplexity and downstream performance with a smaller computational budget and without hybridization of attention. It has already reached State of the art for DNA sequence modeling and is very promising in the field of large language models with Stripped-Hyena-7B. 

Reviewer: ```suggestion
Empirically, Hyena operators are able to significantly shrink the quality gap with attention at scale, reaching similar perplexity and downstream performance with a smaller computational budget and without hybridization of attention. It has already reached State of the art for [DNA sequence modeling](https://arxiv.org/abs/2306.15794) and is very promising in the field of large language models with Stripped-Hyena-7B. 
```
